MSC2176: Update the redaction rules (#2176)
matrix-org/matrix-doc: proposals/2176-update-redaction-rules.md

# MSC2176: Update the redaction rules

The current [redaction
algorithm](https://matrix.org/docs/spec/client_server/r0.5.0#redactions) is now
somewhat dated. This MSC proposes a number of changes to the rules which will
improve the security and reliability of the Matrix protocol.

## Proposal

The following changes will require a new room version, since changes to the
redaction algorithm also change the way that [event
hashes](https://matrix.org/docs/spec/server_server/r0.1.2#calculating-the-reference-hash-for-an-event)
(and hence event IDs) are calculated.

The following *event* keys should be *removed* from the list of those to be
preserved by a redaction:

 * `membership`
 * `prev_state`

> (Note this refers to the *event-level* `membership` property, rather than the

Reviewer: Do we actually have an event-level `membership` property in events? I found nothing of the sort in https://matrix.org/docs/spec/client_server/r0.5.0#event-fields (nor in the S2S spec)

Author: well no, that's the point, and why I believe it should be removed from the list of things to be preserved by a redaction :)

Reviewer: Ah, gotcha, I didn't understand right away what it was supposed to be removed from (nothing to do with the proposal, just my brain malfunctioning)

similarly-named sub-property under the `content` key.)

> Ratinale: neither of the above properties have defined meanings in the Matrix

Reviewer 2: Rationalel

> Ratinale: neither of the above properties have defined meanings in the Matrix

Reviewer 3: Rationale, even :)
```suggestion
Rationale: neither of the above properties have defined meanings in the Matrix
```

protocol, so there is no reason for them to be special-cased in this way.

The following should be added to the list of subkeys of the content property
which should be preserved:

 * `m.room.redaction` should allow the `redacts` key (assuming
   [MSC2174](https://github.com/matrix-org/matrix-doc/pull/2174) is merged).
   Rationale: currently, redacting a redaction can lead to inconsistent results
   among homservers, depending on whether they receive the `m.room.redaction`
   result before or after it is redacted (and therefore may or may not redact
   the original event).

 * `m.room.create` should allow the `room_version` key. Currently, redacting an

>    `m.room.create` event will make the room revert to a v1 room.

Reviewer 4: Also `m.federate` and `predecessor`?

Author: I'm starting to wonder if it would make more sense to just say you can't redact an `m.room.create` event (or rather, that attempting to do so has no effect, so all keys are included in the reference hash).

What does it even mean to redact a create event?

Reviewer 5: I'd heavily vote for not allowing it to be redacted. The creation event only has behaviour if it has `content` (otherwise it's role is simply being the first event in the room), and redacting a state event is not supposed to change the behaviour of the room (at least according to https://github.com/matrix-org/matrix-doc/pull/2032)

Author: done, as of 9e264fe

>  * `m.room.power_levels` should allow the `notifications` key. Rationale:

Reviewer 4: also `invite`?

Author: +1. Never realised that was missing.

Author: fixed by d324cac.

   symmetry with the other `power_levels` settings. (Maybe? See
   https://github.com/matrix-org/matrix-doc/issues/1601#issuecomment-511237744.)

## Potential issues

What if there is spam in sub-properties of the `notifications` property of

> power-levels? Should we not be able to redact it?

Reviewer 3: Sounds exactly like a way to attach a spam dumbbell to a data structure that happened with GPG folks recently. Though much less killing it's no less undesirable.

Reviewer 4: `notifications` doesn't seem any more abusable to me than the other keys in `m.room.power_levels` that we also keep.

Reviewer 5: I'd argue it be kept so that users aren't likely to forget about it when they redact the event (although why would you redact the power levels in the first place?) - if it's the only key excluded, it can be easily missed. However, there is argument that it doesn't help the protocol at all to keep or throw it away, and so we shouldn't bother saving it.

The rationale behind why someone might redact a power level event would be interesting, I think. It could reveal what sorts of keys we should be keeping (from a convenience perspective) and whether it needs to be kept at all.

Author: A solution to spam in the particular case of `notifications` is to specify that only the `room` sub-key be kept. Obviously that's not a general solution though.

Author: > users aren't likely to forget about it when they redact the event 

? You mean client implementors? It would be odd for a user to redact an event manually.

Reviewer 5: I mean more when a user redacts the power level event: if they are redacting to clean up some typos then they might forget to restore their custom `notifications` settings. If they're redacting to reset things to default, then meh, whatever. If they're doing it for some other reason, do we care to try and help by keeping it around?

Reviewer 3: Typos in power levels?.. Seems far-fetched a bit.

similarly-named sub-property under the `content` key.)

Rationale: neither of the above properties have defined meanings in the Matrix
protocol, so there is no reason for them to be special-cased in this way.

The following should be added to the list of subkeys of the content property
which should be preserved:

 * `m.room.redaction` should allow the `redacts` key (assuming
   [MSC2174](https://github.com/matrix-org/matrix-doc/pull/2174) is merged).
   Rationale: currently, redacting a redaction can lead to inconsistent results
   among homservers, depending on whether they receive the `m.room.redaction`
   result before or after it is redacted (and therefore may or may not redact
   the original event).

 * `m.room.create` should allow the `room_version` key. Currently, redacting an
   `m.room.create` event will make the room revert to a v1 room.

 * `m.room.power_levels` should allow:

   * the `invite` key. Rationale: this is required to authenticate
     `m.room.member` events with the `invite` membership. Currently, redacting
     a `power_levels` event will mean that such events cannot be authenticated,
     potentially leading to a split-brain room.

>    * the `notifications` key. Rationale: symmetry with the other `power_levels`

Reviewer 4: Should we also be adding `notifications` too the auth rules so that they're checked in the same way to how the other keys are checked (10.c under https://matrix.org/docs/spec/rooms/v1#authorization-rules)?

Author: probably, but ugh.

Reviewer 5: ftr this is now a dedicated issue: https://github.com/matrix-org/matrix-doc/issues/2198

For simplicity, this can be its own MSC I think.

     settings. (Maybe? See
     https://github.com/matrix-org/matrix-doc/issues/1601#issuecomment-511237744.)

> 

Reviewer 5: `m.room.encryption`'s `algorithm` should be preserved.

Author: I disagree. It doesn't change the behaviour of the room at the server-server level.

Reviewer 3: But if we eventually have two algorithms rather than one how should a client treat an event with redacted algorithm?

Author: > But if we eventually have two algorithms rather than one

I don't think there is a way to end up with two encryption algorithms in the room state?

> how should a client treat `m.room.encryption` with redacted algorithm?

The same way that they should treat a new `m.room.encryption` event which disables encryption: with great suspicion.

Reviewer 5: How are new clients to the room meant to do encryption if they don't know the algorithm? 

Can we please relax the restrictions on what goes into the algorithm? It affects client authors too, and it's ridiculous to exclude something because it doesn't affect the core protocol operation.

Author: > How are new clients to the room meant to do encryption if they don't know the algorithm?

They are not. But I don't understand what adding `algorithm` to the redaction-preservation list buys us, since we have the same problem if you send a new `m.room.encryption` event which disables encryption, which is IMHO a much bigger problem than redactions.

> Can we please relax the restrictions on what goes into the algorithm? It affects client authors too, and it's ridiculous to exclude something because it doesn't affect the core protocol operation.

I assume you mean the *redaction* algorithm here? Sorry, I believe it's important to keep the list as short as possible.

Reviewer 4: I'm inclined to agree with Travis in wanting to preserve the `algorithm` in `m.room.encryption` events, but given that we still have a problem of sending a new `encryption` event, I think it may be better to leave it out for now and tackle both issues in a separate MSC.

Reviewer 6: One of the big motivations for redactions is to allow stripping as much user defined data as possible from an event, without breaking the room in such a way other servers cannot join/interact with it. This was primarily to protect against users adding illegal content to events and servers having no choice but to throw away the room to get rid of it from the DB, though the scope of redactions has increased significantly since.

Author: @erikjohnston: yeeesss... I'm not quite clear what your point is though?

Reviewer 3: I still don't quite understand what a client should do if it loads a room (having no previous state) and receives m.room.encryption with empty algorithm (right away, again, no previous state). Should it consider the room encrypted at all? If yes, which algorithm should it assume?

Author: @KitsuneRal this isn't a question that is limited to redactions, but: I would suggest that a client warns a user that the room state is invalid.

Author: I'm going to resolve this thread; if people feel strongly that I'm making the wrong call on this, then reopen/start a new one.

Reviewer 3: I thought that I sent this comment but actually it got stuck behind the unpressed "Submit review" button (and then GitHub tucked it back where I drafted it. Hope it's placed properly now.

I think I understand where the disconnect is: I'm afraid I don't quite buy the argument that for the sake of reducing the change surface we're restricting redaction exclusions to server interaction pieces. It's the clients ultimately sending and using those events, so redaction should at least somewhat protect basic room integrity from the client perspective as well.

Now the state can be broken in a lot of different ways and (I totally agree here) clients ought to warn users about things that may break their experience; but when it comes to m.room.encryption, there's no good way for a client to handle any readable content of the room once that state is broken.

To reiterate, the case in point is not about repeated m.room.encryption events; it's about a client joining/loading a room where m.room.encryption is already redacted. From the client perspective, this room is irreparably broken along with all of its encrypted messages already in DAG.

Going as far as imagining a m.room.encryption event with obscene/illegal contents for "algorithm" sent to the previously unencrypted room (again - repeated encryption events are not proper and I don't consider them here), in that case the room also gets broken for everyone but no user content is lost - because the room was unencrypted and now it is "encrypted" with non-sensical algorithm. In such case it would be great to have algorithm redacted even though the room has to be thrown out, in order to eliminate improper content from DAG. Ideally such events should not be accepted to DAG in the first place - then the case for the breakage and redaction doesn't emerge.

Author: @KitsuneRal: I think there might be a misunderstanding here.

> there's no good way for a client to handle any readable content of the room once that state is broken
> ...
> this room is irreparably broken along with all of its encrypted messages already in DAG.

`m.room.encryption` is not supposed to define how you interpret received events, so I don't agree that the existing messages are invalidated once `m.room.encryption` is redacted. 

>  again - repeated encryption events are not proper

there is nothing wrong with repeated `m.room.encryption` events?

Reviewer 3: > `m.room.encryption` is not supposed to define how you interpret received events, so I don't agree that the existing messages are invalidated once `m.room.encryption` is redacted.

That's a pretty strong statement. Then what's the purpose of `algorithm` there?

> there is nothing wrong with repeated `m.room.encryption` events?

Except they should be treated with great suspicion because there's no defined semantics for any `m.room.encryption` event except the first one?

Author: > Then what's the purpose of algorithm there?

It indicates to clients how they should encrypt messages in this room.

> there's no defined semantics for any m.room.encryption event except the first one?

I don't think that's true. I can't find anything in the spec about this, but my mental model is that clients should keep a record of rooms which are encrypted (along with the encryption algorithm), and give the user a warning if they send a message in a room which was previously encrypted and is no longer (or if the encryption is 'weaker', for some handwavey definition of weaker).

But certainly it's valid for there to be multiple encryption events in a room, changing minor settings.

Reviewer 5: There's also the problem of clients joining a room which has a redacted encryption event: They won't have the prior knowledge that it was encrypted and will assume it's not.

ftr riot-web does exactly as you describe: it ignores changes to the encryption event it wasn't expecting.

Reviewer 4: I've created https://github.com/matrix-org/matrix-doc/issues/2272 for clarifying the semantics of `m.room.encryption` events.  I'd suggest ignoring `m.room.encryption` for the purposes of this MSC, and discussing in that issue for a future MSC, as I think that it's a larger issue that shouldn't hold this one up.

> # MSC2176: Update the redaction rules

Reviewer 5: Should `m.room.guest_access` preserve `guest_access`? If it doesn't, what does it mean to redact a guest access policy of `can_join`? Ignoring the fact that redacting most state events violates the schemas (`name` for instance is required on the room name event), the guest access event defaults to `forbidden`: are all guests kicked from the room now? How does a remote server even validate that a guest account is a guest account?

I believe this is the only state event which raises questions about authorization: the remainder are already covered by the proposal or the existing algorithm.

Author: remote servers can't distinguish between a guest account and a regular user account, so `m.room.guest_access` isn't involved in the auth checks done on events received over federation (see also: https://matrix.org/docs/spec/rooms/v1#authorization-rules). I'd therefore argue that it's less important to preserve its value than it is for `power_levels` and friends.

To answer your other questions:

> If it doesn't, what does it mean to redact a guest access policy of `can_join`....  the guest access event defaults to forbidden: are all guests kicked from the room now?

according to https://matrix.org/docs/spec/client_server/r0.5.0#id159: yes, although that certainly isn't implemented in synapse currently. (It's also not implemented when the guest access changes due to state resolution rather than due to a new event).

> Ignoring the fact that redacting most state events violates the schemas

yeah; this is a general problem. It begins to interact with validation of events in general (see also: https://github.com/matrix-org/matrix-doc/issues/1646): the current situation is that there is no verification that state events received over federation match the schema, so we have to be prepared to handle events that do not match the schema anyway.

The current [redaction
algorithm](https://matrix.org/docs/spec/client_server/r0.5.0#redactions) is now
somewhat dated. This MSC proposes a number of changes to the rules which will
improve the security and reliability of the Matrix protocol.

## Proposal

The following changes will require a new room version, since changes to the
redaction algorithm also change the way that [event
hashes](https://matrix.org/docs/spec/server_server/r0.1.2#calculating-the-reference-hash-for-an-event)
(and hence event IDs) are calculated.

The following *event* keys should be *removed* from the list of those to be
preserved by a redaction:

 * `membership`
 * `prev_state`

(Note this refers to the *event-level* `membership` property, rather than the
similarly-named sub-property under the `content` key.)

> Rationale: neither of the above properties have defined meanings in the Matrix

Reviewer 5: Historically they did, if someone were to go digging far enough back:
```suggestion
Rationale: neither of the above properties have defined meanings anymore in the Matrix
```

protocol, so there is no reason for them to be special-cased in this way.

> The following should be added to the list of subkeys of the content property

Reviewer 5: ```suggestion
The following are to be added to the list of subkeys of the content property
```

> which should be preserved:

Reviewer 5: ```suggestion
which are preserved:
```

>  * `m.room.create` should preserve *all* content. Rationale: the values in a

Reviewer 5: ```suggestion
 * `m.room.create` preserves *all* content. Rationale: the values in a
```

... and repeat for the rest of the proposal.

>    `create` event are deliberately intented to last the lifetime of the room,

Author: ```suggestion
   `create` event are deliberately intended to last the lifetime of the room,
```

   and if values are redacted, there is no way to add correct settings
   afterwards. It therefore seems non-sensical to allow redaction of a `create`
   event.

 * `m.room.redaction` should allow the `redacts` key (assuming
   [MSC2174](https://github.com/matrix-org/matrix-doc/pull/2174) is merged).
   Rationale: currently, redacting a redaction can lead to inconsistent results
   among homservers, depending on whether they receive the `m.room.redaction`
   result before or after it is redacted (and therefore may or may not redact
   the original event).

>  * `m.room.power_levels` should allow:

Reviewer 5: can we stick some words in here to remind the reader that this list is *adding* to the existing rules and not replacing? After the first two points with some minor context switches, it's easy to forget that this list is additive.

Author: clarification added

   * the `invite` key. Rationale: this is required to authenticate
     `m.room.member` events with the `invite` membership. Currently, redacting
     a `power_levels` event will mean that such events cannot be authenticated,
     potentially leading to a split-brain room.

## Other properties considered for preservation

Currently it is *not* proposed to add these to the list of properties which are
proposed for a redaction:

 * The `notifications` key of `m.room.power_levels`. Unlike the other
   properties in `power_levels`, `notifications` does not play a part in

>    authorising the events in the room graph. Once the `power_levels` are

Reviewer 5: It's true the event graph doesn't care, however the historical context of rooms in clients will be affected by this decision. Currently if one pings the room and redacts the power levels, the client has no choice but to de-highlight the notification. This is less relevant for push notifications as they are already executed prior to the power level event being redacted (it'd be weird and unlikely that someone pings the room then immediately redacts the power levels), however when the user is scrolling back through the room or searching they will have lost highlighting on the event.

tldr: we should preserve it for the purposes of historically accurate highlighting of the room.

Reviewer 3: Is there any difference to redacting, say, a room topic? It also affects user interface, and clients should explicitly process that, and few actually do.

Reviewer 5: The slight difference is that the topic should absolutely be redactable for when someone abuses the text field as a means for sending abuse. The redaction algorithm is also meant to give people the ability to remove harmful content/messages, however whether a message is highlighted or not doesn't really change that.

Author: I'm not sure this is a strong enough argument for me to want to include it in the list. I'll accept lobbying though...

Reviewer 5: My lobbying budget seems to have run out for the quarter :(

If the arguments above aren't convincing enough, there's nothing new I can provide.

   replaced, historical values of the `notifications` property are
   irrelevant. There is therefore no need for it to be protected from
   redactions.

 * The `algorithm` key of `m.room.encryption`. Again, historical values of
   `m.room.encryption` have no effect, and servers do not use the value of the
   property to authenticate events.

>    The effect of redacting an `m.room.redaction` event is much the same as that

Reviewer 3: ```suggestion
   The effect of redacting an `m.room.encryption` event is much the same as that
```

>    of sending a new `m.room.redaction` event with no `algorithm` key. It's

Reviewer 3: ```suggestion
   of sending a new `m.room.encryption` event with no `algorithm` key. It's
```

> # MSC2176: Update the redaction rules

Reviewer 4: i just noticed that the redaction algorithm doesn't care about whether events are state events are not.  Is this something that we need to care about?

Author: possibly. Given the amount of "discussion" which arose from the MSC so far, I've come to the conclusion that having a single MSC to fix all known problems with redaction is misguided - in other words, let's punt that?

Reviewer 4: Fair enough.  I've opened https://github.com/matrix-org/matrix-doc/issues/2308 to track this.

   result before or after it is redacted (and therefore may or may not redact
   the original event).

 * `m.room.power_levels` should allow (in addition to the keys already listed
   in the spec):

   * the `invite` key. Rationale: this is required to authenticate
     `m.room.member` events with the `invite` membership. Currently, redacting
     a `power_levels` event will mean that such events cannot be authenticated,
     potentially leading to a split-brain room.

## Other properties considered for preservation

Currently it is *not* proposed to add these to the list of properties which are
proposed for a redaction:

 * The `notifications` key of `m.room.power_levels`. Unlike the other
   properties in `power_levels`, `notifications` does not play a part in
   authorising the events in the room graph. Once the `power_levels` are
   replaced, historical values of the `notifications` property are
   irrelevant. There is therefore no need for it to be protected from
   redactions.

 * The `algorithm` key of `m.room.encryption`. Again, historical values of
   `m.room.encryption` have no effect, and servers do not use the value of the

>    property to authenticate events.

Reviewer 7: This might be useful for backfilling into E2EE rooms.
